### PR TITLE
Added support for biomes

### LINKED ISSUE #8: Add information on biomes
Following on from MovingBlocks/Terasology#80 we need information on how to create biomes

### PR DESCRIPTION
**Pull request by Adrijaned**
Closes #8 
Adds info on adding biomes based on the rest of the world. The simplest example possible.

### CODE REVIEW

**Comment by Adrijaned** [comment]
Depends on MovingBlocks/Terasology#3560

File: src/main/java/org/terasology/tutorialWorldGenerationLakes/LakesRasterizer.java [hunk 1]
```diff
@@ -21,13 +21,16 @@
...
+import org.terasology.world.generation.Facet;
 import org.terasology.world.generation.Region;
+import org.terasology.world.generation.Requires;
 import org.terasology.world.generation.WorldRasterizerPlugin;
 import org.terasology.world.generation.facets.SeaLevelFacet;
 import org.terasology.world.generation.facets.SurfaceHeightFacet;
 import org.terasology.world.generator.plugin.RegisterPlugin;
 
 @RegisterPlugin
+@Requires({@Facet(SeaLevelFacet.class), @Facet(SurfaceHeightFacet.class)})
```
Comment (Cervator) [inline src/main/java/org/terasology/tutorialWorldGenerationLakes/LakesRasterizer.java:33]: Curiously, is this change related, or just happened to stick out as needed?

**Comment by Cervator** [comment]
Left a single quick line-note :-)

Another general note: I wonder - with us having the extra per-block data system available, what do you think about the idea of turning biome-per-block optional, set up in a module? Not asking for you to implement, just wondering if it would be possible.

New wiki page comment:  Any further ideas/details on why the `implements BiomeRegistrator` bit? Non-intuitively, how come we need it? Any idea if we should change it / any more obvious options?

Engine PR commented posted here because lazy: yep would be good with a little more javadoc 😁 

The Biome enter change event is nice. Could you throw in an entry in the wiki to also cover how you would use such an event, like perhaps a freezing cold biome the causes the player constant environmental damage?

File: src/main/java/org/terasology/tutorialWorldGenerationLakes/LakesRasterizer.java (hunk 1 above)
Comment (Adrijaned) [reply src/main/java/org/terasology/tutorialWorldGenerationLakes/LakesRasterizer.java:33]: This should not change anything as of now, however in theory this should be required on rasterizers/plugins that utilize other facets and do not have them straight away in `@Updates`, as I believe that there is no guarantee in engine API that all the facetProviders will be run before rasterizers, or that they will be run at all even if not explicitly needed (which would for sure provide room for optimization, if only facetProviders for required facets were run)

**Comment by Adrijaned** [comment]
I don't know anything about extra per-block data system, never worked with it before - you mean selecting per-block/per-column biomes or per-block/none-whatsoever biomes?
on the registrator bit - although biomes are and generally should be implemented in enums, someone might feel the need to provide specific additional functionality directly to biome (I doubt it would make any good, but still some people might want to do so). Because of the various ways biomes can be created, they need to be manually registerd, So engine just scans for all the classes implementing the interface and runs them all. If we wanted only biomes implemented using enums, we could just scan for classes implementing `Biome`, register it's values if it is enum and crash otherwise. That would made it simpler.
About the biome event - in this wiki or some else?

**Comment by Cervator** [comment]
Thanks for the comments @Adrijaned :-)

I'll write up a separate issue with details on the extra per-block data thing, no worries there. I've merged the engine PR and would be happy to merge this and approve the GCI task too! But it isn't up for review!

As for the biome change example yep in this wiki with a related code example, should be doable now with the engine PR merged :-)

**Comment by smsunarto** [comment]
Approving the task on GCI and merging this so you can move on with another task @Adrijaned.

Thanks for the work ;)

**Pull request merged**